utils/get_subscription.py

```diff
@@ -0,0 +1,82 @@
+from categories.models import Category,UserSubscription
+
+class get_subscription:
+    should_subscribe = {}
+    subscribe = []
+    unsubscribe = []
+    last_subscribed = {}
+    selected = {}
+    roots = []
+    
+    def __init__(self,save,drop,person,action):
+            self.save = save
+            self.drop = drop
+            self.person = person
+            self.all = save+drop
+            self.action = action
+            self.subscribe = []
+            self.unsubscribe = []
+            self.roots = []
+   
+    #nodes to be selected
+    def select(self):
+        for slug in self.save:
+            self.selected[slug] = True
+        for slug in self.drop:
+            self.selected[slug] = False
+
+    #gets all the roots in all categories
+    def get_roots(self):
+        for slug in self.all:
+            category = Category.objects.get(slug=slug)
+            if category.is_root_node():
+                self.roots.append(category)
+        
+    #checks if this node should be subscribed    
+    def check(self,node):
+        if node.is_leaf_node():
+            if self.selected[node.slug]:
+                self.should_subscribe[node.slug] = True #if node is leaf and was selected node should be subscribed 
+            else:
+                self.should_subscribe[node.slug] = False #if node is leaf and was no selected node shouldn't be subscribed
+        else:
+            children = list(node.get_children())
+            all_selected = True
+            for child in children:
+                self.check(child)
+                if not self.should_subscribe[child.slug]:
+                    all_selected = False
+            
+            self.should_subscribe[node.slug] = False
+            if all_selected:
+                self.should_subscribe[node.slug] = True #if all children need to be subscribed, subscribe this node instead of children
+                for child in children:
+                    self.should_subscribe[child.slug] = False #unsubscribe all children nodes
+        
+    #this function returns which categories to subscribe and unsubscribe
+    def get_should_subscribe(self):
+        #select the nodes to be selected
+        self.select()
+        
+        #get all the roots
+        self.get_roots()
```

Comment:
some variables like root, subscribe, unsubscribe are used only as local variables inside a function, so you need not make them a part of the class i think